### PR TITLE
Support annotations on type parameter declarations

### PR DESCRIPTION
**Pull request by ccantill**
QDox can't parse source code that uses annotations on type parameter declarations (e.g. `List<@NotEmpty String>`). This fixes that.

### CODE REVIEW

**Comment by rfscholte** [comment]
Thanks you for the patch, this would improve QDox.
Looking at the [Java language specification](https://docs.oracle.com/javase/specs/jls/se11/jls11.pdf) it looks like this is fix in the wrong place (you might laready have seen on based on the java comments above it).
A `TypeArgument` can be a `ReferenceType`, which van be a `TypeVariable`, which is a `{Annotation} TypeIdentifier`.

The comments are already there, just not yet implemented:
https://github.com/paul-hammant/qdox/blob/a012901c84b7603855e677b0601b2d1d9e61bdd0/src/grammar/parser.y#L1440-L1457

**Comment by ccantill** [comment]
I redid the fix to adhere to the language spec. But I think this does change how qualified identifiers in the TypeDeclSpecifier are parsed and it may result in the creation of extraneous TypeDefs because there's no easy way to distinguish between packages and classes. It doesn't break any tests, but I'm not 100% sure this is the way it should be fixed.

**Comment by ccantill** [comment]
Any feedback on this PR?

**Comment by rfscholte** [comment]
It still isn't following the described Java Language Specification. What I would expect at least is

 ReferenceType: ArrayType 
              | TypeVariable  
              | ClassOrInterfaceType 
              ;  

Or a proper comment why TypeVariable was implemented in a different way.

**Comment by ccantill** [comment]
TypeVariable (`{Annotation} TypeIdentifier`) is already covered by ClassType (`{Annotation} TypeIdentifier [TypeArguments]`). Added a comment to clarify.

File: src/grammar/parser.y [hunk 1]
```diff
@@ -1665,11 +1665,21 @@ ClassOrInterfaceType: TypeDeclSpecifier
...
+                 | Annotations_opt IDENTIFIER
+                   {
+                     $$ = $2;
+                   }
+                 | ClassOrInterfaceType DOT IDENTIFIER
+                   {
                      $$ = $1.getName() + '.' + $3;
-                   };
+                   }
+                 | ClassOrInterfaceType DOT Annotations_opt IDENTIFIER
```
Comment (rfscholte) [inline src/grammar/parser.y]: The specs don't allow an optional annotation here.
I've tried hard to get this file closer to the Java language specification, this line just doesn't match with the comments above...

Comment (ccantill) [reply src/grammar/parser.y]: That's there to cover this situation (out of the specs):

> PackageName . {Annotation} TypeIdentifier [TypeArguments]
ClassOrInterfaceType . {Annotation} TypeIdentifier [TypeArguments]

File: src/test/java/com/thoughtworks/qdox/GenericsTest.java [hunk 2]
```diff
@@ -139,7 +140,25 @@ public void testShouldUnderstandMethodTypeParameters() {
...
     }
-    
+
+    public void testShouldUnderstandAnnotationsOnTypeParameters() {
+        String source = "" +
+                "public class Bar {" +
+                "    public static Collection<String> foo() { }" +
+                "    public static Collection<java.lang.String> foo2() { }" +
+                "    public static Collection<@Annot String> foo3() { }" +
+                "    public static Collection<java.lang.@Annot String> foo4() { }" +
```
Comment (rfscholte) [inline src/test/java/com/thoughtworks/qdox/GenericsTest.java]: This line looks wrong to me. I assume you `public static Collection<@Annot java.lang.String> foo4() { }`.
If so, I have a patch matching your requirements and the JLS

Comment (ccantill) [reply src/test/java/com/thoughtworks/qdox/GenericsTest.java]: You're right. It does ***look*** wrong. Unfortunately that's how it's supposed to be.

`private Collection<@NotEmpty java.lang.String> test;` does not compile, but
`private Collection<java.lang.@NotEmpty String> test;` does.

That's also what the spec says:

> PackageName . {Annotation} TypeIdentifier [TypeArguments]

Comment (rfscholte) [reply src/test/java/com/thoughtworks/qdox/GenericsTest.java]: https://github.com/paul-hammant/qdox/commit/92e615c8880b2285adb053b01f4f1a84ea715883 contains the proper support for TypeVariable. The other testcase requires support for ClassType (already available as comment, but not implemented and not updated with the latest specs.)
Do you want to give it a try?

Comment (ccantill) [reply src/test/java/com/thoughtworks/qdox/GenericsTest.java]: I think that covers my situation, but it also contains the invalid case of `public static Collection<@Annot java.lang.String> foo4() { }` which should be `public static Collection<java.lang.@Annot String> foo4() { }`

Comment (rfscholte) [reply src/test/java/com/thoughtworks/qdox/GenericsTest.java]: I can't find good examples or explanations, but does `java.lang` belong to `@Annot` or to `String`? Where is this specified?

Comment (rfscholte) [reply src/test/java/com/thoughtworks/qdox/GenericsTest.java]: So it belongs to String. https://github.com/paul-hammant/qdox/commit/8074fba3b02ab1945a1ee49a201829e9a40f55db contains the code to support that too.
I noticed that for JLS11 I should go through the whole file again to get the terms in line again.

**Comment by ccantill** [comment]
Closing this PR as it's already covered by https://github.com/paul-hammant/qdox/compare/TypeVariable

**Pull request closed without merging**